### ISSUE TITLE
torch::autograd::Function should set AutoNonVariableTypeMode when running forward

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#40736]
Currently, it does not, which means you cannot write idiomatic redispatch as:

```
  static torch::autograd::variable_list forward(
      torch::autograd::AutogradContext* ctx,
      torch::autograd::Variable input,
      torch::autograd::Variable rois,
      ...) {
    ctx->saved_data["spatial_scale"] = spatial_scale;
    ...
    // at::AutoNonVariableTypeMode g;
    auto result = roi_align(
        input,
        rois,
        ...
```

You have to add the AutoNonVariableTypeMode guard yourself. This should have been taken care of in Function.

**Comment by ezyang** [comment]
Maybe this is intended behavior though: with the current behavior, you can write a complete forward-backward pass reentrantly inside of forward; but then simply define the double backwards formula directly. So maybe this is NOTABUG.

**Comment by ezyang** [comment]
Some extra context from private correspondence:

> full disclosure I'm not parsing the use case described in your NOTABUG pitch, but... it does feel like a bug from the obvious layman's POV

Basically, in Python, torch.autograd.Function doesn't magically cause you to bypass dispatch to autograd. Therefore, C++ should behave the same way.

> Sure, didn't realize that. why is that? (in python)

Well, not too long ago, there wasn't any such thing as "bypassing dispatch".

**Comment by mcarilli** [comment]
Thanks for following up!
> Basically, in Python, torch.autograd.Function doesn't magically cause you to bypass dispatch to autograd.

hmm i don't think I understand what you mean.  Here's what I observe on a recent master:
```python
import torch

class F(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input):
        print(torch.is_grad_enabled())
        return input
    @staticmethod
    def backward(ctx, grad):
        print(torch.is_grad_enabled())
        return grad

F.apply(torch.ones(1, requires_grad=True)).backward()
# prints
# False => autograd is disabled in F.forward()
# False => autograd is disabled in F.backward()

F.apply(torch.ones(1, requires_grad=True)).backward(create_graph=True)
# prints
# False => autograd is disabled in F.forward()
# True => autograd enabled in F.backward()
```

**Comment by ezyang** [comment]
`is_grad_enabled` is a different concept from AutoNonVariableTypeMode. If you did this test in C++ you would find that Python and C++ agree :)

Part of the confusion here is that, from a user persective, AutoNonVariableTypeMode doesn't make sense; it's a very operational guard whose semantics depends on how the dispatcher works. We have some ongoing work to unify the two concepts, but we're not there yet.

**Comment by ezyang** [comment]
I'm going to reopen this issue for visibility.

**Comment by mcarilli** [comment]
Looks like `is_grad_enabled()` maps to some [standalone thread local flag](https://github.com/pytorch/pytorch/blob/179dbd4f25288a098099e806b16098468e6f0b2f/aten/src/ATen/core/grad_mode.cpp#L13) separate from any dispatch keys.  This still has an effect:  it [prevents autograd history recording](https://github.com/pytorch/pytorch/blob/179dbd4f25288a098099e806b16098468e6f0b2f/torch/csrc/autograd/functions/utils.h#L45-L48) if a call happens to route through a `VariableType*.cpp` function.  However, it _does not prevent_ calls from dispatching into `VariableType*.cpp` functions to begin with.  That dispatch is triggered by `DispatchKey::autograd`, which is [always enabled](https://github.com/pytorch/pytorch/blob/179dbd4f25288a098099e806b16098468e6f0b2f/aten/src/ATen/core/dispatch/DispatchKeyExtractor.h#L18) unless explicitly excluded.

(I'm not sure if `DispatchKey::autograd` is also enabled in the `key_set_` of Tensors that require grad.  Trying to figure out where during tensor construction `key_set_` gets populated.)

Meanwhile `AutoNonVariableTypeMode` is what [explicitly excludes DispatchKey::Autograd](https://github.com/pytorch/pytorch/blob/179dbd4f25288a098099e806b16098468e6f0b2f/aten/src/ATen/core/LegacyTypeDispatch.h#L50).

In light of above, the following comment [above `AutoNonVariableTypeMode`](https://github.com/pytorch/pytorch/blob/8e0714a60de4a7ff0c1b131ba8fa25868524907c/aten/src/ATen/core/LegacyTypeDispatch.h#L39-L44) makes sense
> // TODO: Since `torch::NoGradGuard` serves almost the same purpose in libtorch,
// we should merge these two thread-local guards.  However, NoGradGuard does
// something subtly different: it turns off gradient recording, but DOES NOT
// skip VariableType implementation (as we still might need to profile or
// trace).  To unify the two, we would first have to move profiling and tracing
// out of VariableType.

and I understand why the `AutoNonVariableType` guard in eg. [ROIAlignFunction::forward](https://github.com/pytorch/vision/blob/766721b1dfd7a92130146a549c4fcca15cc069b2/torchvision/csrc/ROIAlign.h#L86) is helpful.

Mostly writing this out so I can bookmark it for medium-term guidance while working with torchvision.  Thanks for the explanations!  (also, please correct me if anything is wildly wrong)

**Comment by ezyang** [comment]
Yes this is correct! And you indeed get an infinite loop if you forget to add the guard. This is something that's very easy to get wrong and I suspect more people will run into it.
